diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 4c06409..c023910 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -234,6 +234,11 @@ IF (PAHO_WITH_SSL)
             TARGET_LINK_LIBRARIES(${TARGET}
                 PUBLIC
                     OpenSSL::SSL OpenSSL::Crypto ${LIBS_SYSTEM})
+            IF (WIN32)
+                TARGET_LINK_LIBRARIES(${TARGET}
+                    PUBLIC
+                        bcrypt)
+            ENDIF()
         ENDFOREACH()
         INSTALL(TARGETS paho-mqtt3cs paho-mqtt3as
             EXPORT eclipse-paho-mqtt-cTargets
@@ -300,6 +305,11 @@ IF (PAHO_WITH_SSL)
             TARGET_LINK_LIBRARIES(${TARGET}
                 PUBLIC
                     OpenSSL::SSL OpenSSL::Crypto ${LIBS_SYSTEM})
+            IF (WIN32)
+                TARGET_LINK_LIBRARIES(${TARGET}
+                    PUBLIC
+                        bcrypt)
+            ENDIF()
         ENDFOREACH()
     ENDIF()
 ENDIF()
diff --git a/src/SSLSocket.c b/src/SSLSocket.c
index fd80c72..1dbd03b 100644
--- a/src/SSLSocket.c
+++ b/src/SSLSocket.c
@@ -593,7 +593,7 @@ int SSLSocket_createContext(networkHandles* net, MQTTClient_SSLOptions* opts)
 		}
 	}
 
-#if (OPENSSL_VERSION_NUMBER >= 0x10100000L)
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L && !defined(LIBRESSL_VERSION_NUMBER))
 	SSL_CTX_set_security_level(net->ctx, 1);
 #endif
 
